src/codeeditor/editor.ts

```diff
@@ -0,0 +1,804 @@
+
+    /**
+     * The bottom coordinate value.
+     */
+    readonly bottom: number;
+  }
+
+  /**
+   * Configuration options for the editor.
+   */
+  export
+  interface IConfiguration extends IDisposable {
+    /**
+     * A signal emitted when the configuration changes.
+     */
+    changed: ISignal<IConfiguration, string>;
+
+    /**
+     * Whether line numbers should be displayed. Defaults to false.
+     */
+    lineNumbers: boolean;
+
+    /**
+     * Set to false for horizontal scrolling. Defaults to true.
+     */
+    wordWrap: boolean;
+
+    /**
+     * Whether the editor is read-only.  Defaults to false.
+     */
+    readOnly: boolean;
+
+    /**
+     * The number of spaces a tab is considered equal to. Defaults to 4.
+     */
+    tabSize: number;
+
+    /**
+     * Translate tabs to spaces. Defaults to true.
+     */
+    tabsToSpaces: boolean;
+
+    /**
+     * Set to false to stop auto pairing quotes, brackets etc.
+     * Defaults to true.
+     */
+    autoMatchEnabled: boolean;
+
+    /**
+     * Set to false to disable highlighting the brackets surrounding the
+     * cursor. Defaults to true.
+     */
+    matchBrackets: boolean;
+
+    /**
+     * Set to false to disable scrolling past the end of the buffer.
+     * Defaults to false.
+     */
+    scrollPastEnd: boolean;
+  }
+
+  /**
+   * A text selection.
+   */
+  export
+  interface ITextSelection {
+    /**
+     * The index to the first character in the current selection.
+     */
+    start: number;
+
+    /**
+     * The index to the last character in the current selection.
+     */
+    end: number;
+
+    /**
+     * The direction in which selection occurred.
+     */
+    direction: 'forward' | 'backward' | 'none';
+  }
+
+  /**
+   * An editor model.
+   */
+  export
+  interface IModel extends IDisposable {
+    /**
+     * A signal emitted when the value changes.
+     */
+    valueChanged: ISignal<IModel, IChangedArgs<string>>;
+
+    /**
+     * A signal emitted when a property changes.
+     */
+    propertyChanged: ISignal<IModel, IChangedArgs<any>>;
+
+    /**
+     * The text stored in the model.
+     */
+    value: string;  // TODO: this should be an iobservablestring.
+
+    /**
+     * The uri associated with the model.
+     */
+    uri: string;
+
+    /**
+     * A mime type of the model.
+     */
+    mimeType: string;
+
+    /**
+     * The cursor position.
+     */
+    cursor: ICursorPosition;
+
+    /**
+     * Get the number of lines in the model.
+     */
+    readonly lineCount: number;
+
+    /**
+     * The currently selected code.
+     */
+    selection: ITextSelection;
+
+    /**
+     * Returns the content for the given line number.
+     */
+    getLine(line: number): string;
+
+    /**
+     * Find an offset for the given position.
+     */
+    getOffsetAt(position: ICursorPosition): number;
+
+    /**
+     * Find a position fot the given offset.
+     */
+    getPositionAt(offset: number): ICursorPosition;
+
+    /**
```

Comment:
Are `undo`, `redo` and `clearHistory` CodeMirror specific? Should they be exposed as common for all editors? Is there another way to handle it?